homeassistant/util/dt.py

```diff
@@ -337,11 +344,32 @@ def _lower_bound(arr: list[int], cmp: int) -> int | None:
     # For example: if triggering on 2:30 and now is 28.10.2018 2:30 (in DST)
     # we should trigger next on 28.10.2018 2:30 (out of DST), but our
     # algorithm above would produce 29.10.2018 2:30 (out of DST)
-    if tz.datetime_ambiguous(now):
+    if datetime_ambiguous(now):
         check_result = find_next_time_expression_time(
             now + _dst_offset_diff(now), seconds, minutes, hours
         )
-        if tz.datetime_ambiguous(check_result):
+        if datetime_ambiguous(check_result):
             return check_result
 
     return result
+
+
+def datetime_exists(dattim: dt.datetime) -> bool:
+    """Check if a datetime exists."""
+    assert dattim.tzinfo is not None
+    original_tzinfo = dattim.tzinfo
+    # Check if we can round trip to UTC
+    return dattim.replace(tzinfo=None) == dattim.astimezone(UTC).astimezone(
```

Comment:
You should be able to skip the `.replace(tzinfo=None)` on both sides of this equation, since intra-zone comparisons only look at the naïve portion of the datetime anyway. Since each `.replace` requires constructing one datetime, and datetime is somewhat expensive to construct (fast paths on my laptop are around 100ns, though weirdly `dt.replace(tzinfo=None)` is consistently measuring 1 *micro*second!!!), it's probably best to avoid the extra work.